ioBroker/ioBroker.node-red
# Replace HTTP Basic Authentication by Node-RED Login Page

Hi @GermanBluefox , @Apollon77,
given the issues #98 , #58 I thought I'd give my first PR for this repository a try.

## Implemented changes
* replaced setting `httpAdminAuth` by `adminAuth` in file settings.js
* Adapted setting replacement code in file main.js
* Changed password hashing from MD5 to bcrypt.js in file index_m.html
* Added bcrypt.min.js and linked it in index_m.html

Implementation based on Security guidance by Node-RED: https://nodered.org/docs/user-guide/runtime/securing-node-red#usernamepassword-based-authentication

## Testing done so far
* I did various tests by installing the adapter into ioBroker from my Github fork (`npm install https://github.com/jwiesel/ioBroker.node-red/tarball/master`) as indicated in the ioBroker documentation: https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/adapterdev.md#deploying
 
## Test Instructions
* Install the adapter in your ioBroker installation
* Set or replace your password in the Node-RED instance settings. - Important: As you might still have an MD5 based password stored, that has to be replaced by the bcrypt based password. After saving the settings, wait a few seconds until Node-RED is up and running again
* Open node-red --> now the node-red "login page" should load instead of the HTTP Basic Authentication popup. Login with your newly set username + password
* Logout out of Node-RED again
* Go to the instance settings and remove username + password. Wait a few seconds until Node-RED is up and running again
* Open node-red / F5 to reload the page again. Now you should directly see your flows without the need to login

## Observations
* The bcrypt algorithm seems to need a bit longer than expected by ioBroker during save operation of the instance settings. I sometimes get a message after clicking on "Save", that I have unsaved changes which I might lose, when leaving the page. I always click on OK and it is saved anyhow. Maybe someone has an idea to to avoid the popup

## Further improvement suggestions
* If you agree to merge this PR, I could remove the MD5 code beforehand from index_m.html to avoid unused overhead code.
